### PR TITLE
Fix / improve window behaviour: restore pre-fullscreen/pre-scratch window size/position; better manage windows that start fullscreen

### LINKED ISSUE #638: Full screen Wayland apps show black screen and top left window corner near center of screen
**Describe the bug**
I tried to watch a movie in one of streaming services and I pressed F11 in firefox to go to full screen. I saw black screen and window's top left corner was near the middle of screen. Similar behaviour was when I pressed win+shift+f. Same behaviour with kitty (terminal emilator) and other apps.

The other strange behaviour - it's happening only to windows that are last on the right. For second (third...) to last windows maximization shows 1 frame with problem, but then it fixes itself. My Signal app with ozone/wayland enabled is not counted.

**To Reproduce**
1. Put window to maximize as the last window
2. Press win+shift+f to maximize window to full screen

**Expected behavior**
Full screen windows should not have top left corners near middle of the screen.

I use Arch Linux and newest PaperWM from extensions.gnome.org

### LINKED ISSUE #739: Fullscreening a window and then unfullscreening change the window size (from original)
**Describe the bug**
Fullscreening a window and then unfullscreening change the window size (from original)

**To Reproduce**
1. fullscreen a window
2. unfullscreen a window

Do multiple times and the original window's size has changed.

**Expected behavior**
Window maintains it's original size after fullscreening. 

**System information:**
Please execute `./gather-system-info.sh` in you PaperWM clone and paste the output below.

```
Distribution: EndeavourOS
GNOME Shell 45.2
Display server: Wayland
PaperWM branch/tag: override-switch-windows
PaperWM commit: f769173bf3d338b663e5a70e51f60b1f60d04c44
Enabled extensions:
- Vitals@CoreCoding.com
- caffeine@patapon.info
- just-perfection-desktop@just-perfection
- appindicatorsupport@rgcjonas.gmail.com
- launch-new-instance@gnome-shell-extensions.gcampax.github.com
- switcher@landau.fi
- hibernate-status@dromi
- paperwm@paperwm.github.com

```

### PR DESCRIPTION
**Pull request by jtaala**
Fixes #739, and partially fixes #638.

This PR includes improvements to window fullscreening behaviours:

1. better handle application windows that start as fullscreen windows;
2. restores window size and positions after un-fullscreening;
3. scratching a window now properly activates that window (previously is wasn't properly activated);
4. restores window size and positions after when scratching a window;

(1) is part of the fix required for #638 (the other part is an upstream issue that will be fixed in mutter with patch [3401](https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/3401) - which has been merged but not released yet).  

These changes fix issues for me in [kodi](https://github.com/paperwm/PaperWM/issues/638#issuecomment-1826662458) - but not the osu! issues (which requires also requires mutter fixes as mentioned above).

### CODE REVIEW

File: tiling.js [hunk 1]
```diff
@@ -380,6 +380,15 @@ export class Space extends Array {
...
         });
 
+        // after layout is complete, remove any residual fullscreenOnLayout
+        this.signals.connect(this, "layout", () => {
+            this.getWindows().forEach(w => {
+                if (w.fullscreenOnLayout) {
+                    delete w.fullscreenOnLayout;
+                }
+            });
+        });
```
Comment (Lythenas) [inline tiling.js]: Why is this here and not at the end of `layout`?

Comment (jtaala) [reply tiling.js]: Good question!

No reason really, other than using the "layout" signal (we don't use it so I thought "why not?").

It probably should be in `layout` and at as a `MetaLater`.

This PR still has a few issues that I'm working though (mainly, with `kodi` on unfullscreening, it doesn't resize itself... it's more a `kodi` thing but I'd like to address).  I'll set this as a draft for now.

File: tiling.js [hunk 2]
```diff
@@ -698,17 +706,24 @@ export class Space extends Array {
             this.moveDone();
         }
 
+        // if only one window on space, then centre it
```
Comment (Lythenas) [inline tiling.js]:
```suggestion
        // if only one window on space, then center it
```

**Review by Lythenas — CHANGES_REQUESTED** [review]
As far as I can tell code seems fine.

I noticed one thing: After starting chromium in fullscreen (`chromium --start-fullscreen`) and unfullscreening (F11), the topbar is on top of the window. Focus another monitor and focusing back fixes this. I think there is a layout call missing after unfullscreening the window or unhiding the topbar. Might also be a race condition (so layout is called to early).

Starting kodi in fullscreen works fine now.

> restores window size and positions after when scratching/un-scratching a window;

Restoring size after unscratching works, but restoring position does not work. Had two windows on workspace firefox full width, terminator smaller on the left, scratched terminator, unscrated terminator, terminator is put on the right. To be honest I don't see this as very important.

**Comment by jtaala** [comment]
> but restoring position does not work.

Ah, no this wasn't the intent - I meant on scratch, but maybe that always worked (restoring position of scratched window).

**Comment by jtaala** [comment]
> After starting chromium in fullscreen (`chromium --start-fullscreen`) and unfullscreening (F11),

I can't seem to reproduce this.  Let me know if I'm doing anything wrong here:

https://github.com/paperwm/PaperWM/assets/30424662/809f62c3-5ccb-45fd-a55c-b41c045abd21

**Comment by jtaala** [comment]
> I noticed one thing: After starting chromium in fullscreen (`chromium --start-fullscreen`) and unfullscreening (F11), the topbar is on top of the window. Focus another monitor and focusing back fixes this. I think there is a layout call missing after unfullscreening the window or unhiding the topbar. Might also be a race condition (so layout is called to early).

Try pulling the latest.  I got sick of all of the corner cases for the `spaceTopBarElements` method so I reviewed/rewrote it (to be much more self-contained).  It annoyed me since after unfullscreening the topbar style would often be wrong.

**Comment by jtaala** [comment]
Let me know if you're still seeing the chromium issue as there's something else I can try (essentially calling `fixTopbar` in the unfullscreening callback).

**Comment by jtaala** [comment]
Thanks @Lythenas,

I've made a few changes (some that impact topbar showing).

Let me know if you're still seeing the chromium issue after pulling the latest on this PR as there's something else I can try (essentially calling `fixTopbar` in the unfullscreening callback).

**Comment by Rutherther** [comment]
This solves the issue for me when I am moving the window itself between monitors (Move active window to the {direction} monitor) - before when moved the window twice I got the black screen part. Unfortunately it doesn't solve it when moving the whole workspace (Swap workspace with monitor {direction}). Then I get the issue every time - the same as before.

Also with that moving between monitors, I tried moving the window very fast, and sometimes I got the black screen part, then moved it one more time and it went away. I wouldn't worry about this since as jtalaa said, it's more of a workaround, and also, moving windows this fast is not really something one would do regularly, at least I would guess that's the case.

**Comment by jtaala** [comment]
> Unfortunately it doesn't solve it when moving the whole workspace (Swap workspace with monitor {direction}). Then I get the issue every time - the same as before.

Thanks @Rutherther - I can reproduce this with firefox (swapping) so will see if anything can be done for this this one.

**Comment by Lythenas** [comment]
> > After starting chromium in fullscreen (`chromium --start-fullscreen`) and unfullscreening (F11),
> 
> I can't seem to reproduce this. Let me know if I'm doing anything wrong here:

Ah, I noticed that it seems to be a problem because I had Fullscreen Avoider enabled. After disabling that it works fine. So I think you can ignore that. If it is still an issue I will look at it in the other PR.

After disabling this is how it looks for me:
[Screencast from 2024-01-01 12-39-20.webm](https://github.com/paperwm/PaperWM/assets/32217236/47710693-69b2-4914-a978-7e0961e6917f)

Is there something we can do about the topbar backdrop that flashes up after launching chromium?

**Comment by jtaala** [comment]
@Rutherther - do a `git pull` on this PR, logout/login and give it another go (added a temp workaround for swap monitors that seems to be working for me.

**Comment by jtaala** [comment]
> Is there something we can do about the topbar backdrop that flashes up after launching chromium?

Yeah, there's still a few issues with topbar control here that I need to have a look at.

**Comment by jtaala** [comment]
> Is there something we can do about the topbar backdrop that flashes up after launching chromium?

PR now disables window position bar for space when fullscreen window is focused (or opened).

**Comment by Rutherther** [comment]
> @Rutherther - do a `git pull` on this PR, logout/login and give it another go (added a temp workaround for swap monitors that seems to be working for me.

Perfect! Yes, thank you. This PR also fixes #722, although when opening an image, at first there is the bug, and then it seems to be corrected somehow, (it's quite quick, but noticeable)

**Comment by Lythenas** [comment]
> Perfect! Yes, thank you. This PR also fixes #722, although when opening an image, at first there is the bug, and then it seems to be corrected somehow, (it's quite quick, but noticeable)

I also noticed this. Sometimes with Terminator and with Firefox. But it's not at all consistent...

**Comment by Rutherther** [comment]
One more thing about swapping workspaces. I still get the bug when I trigger swap from a workspace without fullscreen window. Also when I trigger it like described, and unfullscreen, the position is wrong. This happens to me with Firefox.

**Comment by Rutherther** [comment]
> > Perfect! Yes, thank you. This PR also fixes #722, although when opening an image, at first there is the bug, and then it seems to be corrected somehow, (it's quite quick, but noticeable)
> 
> I also noticed this. Sometimes with Terminator and with Firefox. But it's not at all consistent...

Yes, I get this unconsistently with Firefox. With Telegram images it's consistent, and longer than the one with Firefox.

**Comment by jtaala** [comment]
Yes, unfortunately firefox is highly susceptible to this issue (as are other apps like osu!, kodi, remminna etc).  

We're not going to be able to workaround this bug entirely from occurring in firefox (or osu!) until mutter is updated/released  patch 3401.

**Comment by Rutherther** [comment]
I understand that. I am giving these issues just to let you know where another possibilities to fix lay. Of course if you think this is more than the scope of this fix, I don't want you to implement it.

I have actually found out that it's not just swapping the workspace with fullscreened window, it's even swapping only two unrelated workspaces on other two monitors, without touching the workspace with fullscreen window. So this adds a new trigger for the bug somewhere, as this wasn't happening prior to this pull request

**Comment by jtaala** [comment]
No probs @Rutherther.

So, I believe I've finally found the main trigger for the blackscreen bug and I'm now working on an implementation to avoid the trigger.  It's actually related to https://github.com/paperwm/PaperWM/pull/650 and triggered by fullscreening a window and then moving repositioning it with paperwm's `move_to` method (which was needed to fix the issue in #650 (otherwise when fullscreening a window close to a monitor edge - it would fullscreen on the other monitor).

A better approach I'm working on is to intercept a fullscreen call and immediately unfullscreen/reject it, then move the window to a safe position, then actually fullscreen the window... that's the theory anyway.

It'll take me a couple of days to implement changes here but preliminary tests look good (which is just me opening firefox and spamming fullscreen on it looking for the blackscreen bug).

I'll put this PR in draft for now while I work on this.

**Review by Thesola10 — APPROVED** [review]

**Review by Thesola10 — DISMISSED** [review]
These changes break YouTube full screen on my system:

image

This is a regression from the version I previously approved.

Kodi fullscreen issue also came back.

**Comment by jtaala** [comment]
> These changes break YouTube full screen on my system:
> 
> image
> 
> This is a regression from the version I previously approved.
> 
> Kodi fullscreen issue also came back.

Whoah - slow down - still a WIP (hence this is a draft now).  No where near finished here.

i.e. see last commit message WIP.

**Comment by Thesola10** [comment]
originally meant to dismiss my review, couldn't find the option :p

**Comment by jtaala** [comment]
The other option is I can move this new approach I'm working on to a new branch and release the commit you approved?

**Comment by Thesola10** [comment]
> The other option is I can move this new approach I'm working on to a new branch and release the commit you approved?

honestly sounds like a plan

**Comment by jtaala** [comment]
> > The other option is I can move this new approach I'm working on to a new branch and release the commit you approved?
> 
> honestly sounds like a plan

Yeah, especially since this new approach might turn out to have more issues than the previous approach...

**Comment by jtaala** [comment]
@Thesola10 - do have the commit number you approved?

**Comment by Thesola10** [comment]
I had `c0153754` tagged as working before pulling updates today

**Comment by jtaala** [comment]
Okay, I'm going to reset this PR branch back to that commit.

**Comment by jtaala** [comment]
Hey @Thesola10, can you pull and test the latest on this branch.  A few small changes here:
- behaviour change: if you have a fullscreen window and select another tiled window (on the space) - unfullscreen the fullscreened window.  This feels more natural to me and allows us to fix the following multimonitor bugs:
- remove flash of topbar spacing when moving mouse to another monitor;
-  stop window position bar from showing on monitor you left when it has a fullscreen window.

**Review by Thesola10 — APPROVED** [review]
feels pretty nice, fixes fullscreen issues

**Comment by Lythenas** [comment]
> behaviour change: if you have a fullscreen window and select another tiled window (on the space) - unfullscreen the fullscreened window. This feels more natural to me and allows us to fix the following multimonitor bugs:

I think that could cause problems. I guess mostly with games or other things that are supposed to be always fullscreen. But let me test it a bit.

**Comment by Lythenas** [comment]
Actually immediately found some^^

Normally when this game is fullscreen and you focusing another window (on the current or other monitor) the game minimizes). You have to Alt+Tab back to it again.

With these changes after navigating to another window on the same monitor, it looks like this:
Screenshot from 2024-01-04 18-33-36

The game still shows on top of other windows but you click through it to the window that is below. You can still get back to the game via Alt+Tab. Focusing another monitor while the game is maximized still minimizes it.

This means if you don't have a second monitor you can't tab out of the game anymore.

Another game (MTG Arena) is basically the same, except it takes up the entire screen and you just click through it but can't see anything.

Another game (Warframe) does not become fullscreen again after focusing away and back to it. You have to Alt+Enter twice (once to go to windowed mode and the second time to go to fullscreen mode again). I think that is implemented by the game. This game also does not lose focus when navigating to another monitor. So focusing another window on the same monitor is the only way to tab out of the game in PaperWM.

So I'm against the last couple of commits. Games make enough problems on Linux, we should try to avoid introducing any more^^

I think I was also on commit c0153754 before and that worked fine.

Regarding the topbar behavior we can probably also change that without unfullscreening a window. E.g. we could consider a space fullscreen only if the *active* window is fullscreen (instead of any window). In #707 I also try to fix some of these issues.

**Comment by Lythenas** [comment]
Also I noticed with the last commits navigating workspaces via Super+PageUp/Down does not complete the navigation. You are stuck in the zoomed out animation until you click a window to focus it.

Edit: Actually also the window navigation is the same. The window preview does not close until you click on a window. I saw this error in the logs. But it's not always generated:

```
gnome-shell[14942]: JS ERROR: TypeError: this.signals is null
                                            destroy@file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/navigator.js:254:9
                                            dismissDispatcher@file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/navigator.js:468:20
                                            _finish@file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/navigator.js:228:26
                                            finishDispatching@file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/navigator.js:454:24
                                            select@file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/stackoverlay.js:211:19
                                            ClickOverlay/<@file:///home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/stackoverlay.js:131:18
                                            @resource:///org/gnome/shell/ui/init.js:21:20
```

**Comment by jtaala** [comment]
> Actually immediately found some^^

Into the Breach!

Thanks @Lythenas for the further testing there.  I've reverted the unfullscreening approach.  And one of the fixes that relied on it.

Still have the small "bobble" issue, it was just a check  on layout.  Will see if there's something quick that can be done here to resolve that one too, otherwise will leave that one for later.

**Comment by jtaala** [comment]
Hey all, I believe I've fixed the "bobble" issue when moving mouse to another monitor when a fullscreen window is active.

Give this one a go.

**Review by Lythenas — APPROVED** [review]
Works fine now.

**Pull request merged**